cipher/ascon/ascon.go

```diff
@@ -0,0 +1,260 @@
+// Package ascon provides a light-weight AEAD cipher.
+//
+// This packges implements the AEAD ciphers Ascon128 and Ascon128a as specified
+// in https://ascon.iaik.tugraz.at/index.html
+package ascon
+
+import (
+	"bytes"
+	"crypto/cipher"
+	"encoding/binary"
+	"errors"
+	"math/bits"
+)
+
+const (
+	KeySize   = 16
+	NonceSize = 16
+	TagSize   = KeySize
+)
+
+type Mode int
+
+const (
+	Ascon128 Mode = iota + 1
+	Ascon128a
+)
+
+const (
+	permA     = 12
+	permB     = 6 // 6 for Ascon128, or 8 for Ascon128a
+	blockSize = 8 // 8 for Ascon128, or 16 for Ascon128a
+	ivSize    = 8
+	stateSize = ivSize + KeySize + NonceSize
+)
+
+var (
+	roundConst = [12]uint64{0xf0, 0xe1, 0xd2, 0xc3, 0xb4, 0xa5, 0x96, 0x87, 0x78, 0x69, 0x5a, 0x4b}
+	subs       = [32]int{
+		0x04, 0x0b, 0x1f, 0x14, 0x1a, 0x15, 0x09, 0x02,
+		0x1b, 0x05, 0x08, 0x12, 0x1d, 0x03, 0x06, 0x1c,
+		0x1e, 0x13, 0x07, 0x0e, 0x00, 0x0d, 0x11, 0x18,
+		0x10, 0x0c, 0x01, 0x19, 0x16, 0x0a, 0x0f, 0x17,
+	}
+)
+
+type Cipher struct {
+	state [stateSize]byte
+	key   [KeySize]byte
+	mode  Mode
+}
+
+// New returns a Cipher struct implementing the cipher.AEAD interface. Mode is
+// one of Ascon128 or Ascon128a.
+func New(key []byte, m Mode) (*Cipher, error) {
+	if len(key) != KeySize {
+		return nil, ErrKeySize
+	}
+	if !(m == Ascon128 || m == Ascon128a) {
+		return nil, ErrMode
+	}
+	c := new(Cipher)
+	c.mode = m
+	copy(c.key[:], key)
+	var _ cipher.AEAD = c
```

Comment:
^ why?

Reply:
it's a compile-time check that `c` implements the AEAD interface.

Reply:
It's a bit out of place here — better to make it part of a test.

Reply:
done